docs/DesignDocs/ExtendedEnclaveInitializationData.md

```diff
@@ -0,0 +1,168 @@
+    uint64_t entry_point;    /* entry point of the image */
+    uint64_t data_size;      /* size of application EEID */
+    uint8_t data[];          /* actual application EEID */
+} oe_eeid_t;
+```
+
+Once an enclave has been started with EEID, it can use `oe_get_report` as usual to obtain evidence, except that this evidence will not be verifiable by enclaves built without EEID support, e.g. the Intel SGX SDK.
+However, enclaves built with OE can appraise this evidence, using `oe_verify_report` as usual, regardless of whether they themselve use EEID, as long as their version of OE supports EEID.
+
+Specification
+-------------
+
+![Design Overview](eeid.png "EEID Design Overview")
+
+The changes introduced by EEID are mostly internal to the enclave initialization function (`oe_create_enclave_eeid`) and the attestation functions (`oe_get_report`/`oe_verify_report`).
+
+First, before an enclave image with EEID is loaded, we patch the size settings from the signed image with instance-specific settings.
+This alters the computation of the enclave hash (e.g. `MRENCLAVE` on SGX) as the sequence of protected memory pages (e.g. via `EADD/EEXTEND`) is modified with EEID after signing. Thus, we save the state of the hash computation before adding these pages, as it will be needed to verify EEID evidence. This also requires minor additions to `oesign`, such that the extended metadata (base image `hash_state`, `vaddr`, and `sigstruct`) can be dumped from signed enclave images, to enable the prediction of hashes before an enclave is started.
+
+```C
+    if (eeid)
+    {
+        // Check that size settings are zero as an indication that the
+        // image was intended to be used with EEID.
+        if (props.header.size_settings.num_heap_pages != 0 ||
+            props.header.size_settings.num_stack_pages != 0 ||
+            props.header.size_settings.num_tcs != 0)
+            OE_RAISE(OE_INVALID_PARAMETER);
+
+        props.header.size_settings.num_heap_pages =
+            eeid->size_settings.num_heap_pages;
+        props.header.size_settings.num_stack_pages =
+            eeid->size_settings.num_stack_pages;
+        props.header.size_settings.num_tcs = eeid->size_settings.num_tcs;
+
+        // Patch ELF symbols
+        elf64_sym_t sym_props = {0};
+        const elf64_t* eimg = &oeimage.u.elf.elf;
+        if (elf64_find_symbol_by_name(
+                eimg, "oe_enclave_properties_sgx", &sym_props) == 0)
+        {
+            uint64_t* sym_props_addr = NULL;
+            sym_props_addr =
+                (uint64_t*)(oeimage.image_base + sym_props.st_value);
+            oe_sgx_enclave_properties_t* p =
+                (oe_sgx_enclave_properties_t*)sym_props_addr;
+            p->header.size_settings = props.header.size_settings;
+        }
+    }
+```
+
+Then, after the enclave image is loaded but before it is initialized, we add the additional memory pages containing the contents of the `oe_eeid_t` struct:
+
```

Comment:
The new pages are measured and thus the final mrenclave is different, so there is no reason for `enclave_initialize` to fail.

Reply:
The more detailed answer probably should be "the extra pages are both measured in software (`oe_sgx_measure_load_enclave_data`)  and hardware (`enclave_load_data`), so `enclave_initialize` succeeds because both software and hardware agree on the same final mrenclave value."

Reply:
Yes, they both measure the same pages.